**Astroinformatica I (Semester 1 2025)**
# Tutorial Session 5: Python Introduction

*N. Hernitschek*


---
## Contents
* [Python Review](#first-bullet)
* [Packages, Libraries, Modules](#second-bullet)
* [Coding Exercise](#third-bullet)
* [Summary](#nineth-bullet)


## 1. Python Review <a class="anchor" id="first-bullet"></a>

Before we continue with specific coding exercises, please review the code examples we saw in the lecture.
Reviewing code examples should include:
* running the code
* modifying the code to get a general understanding
* trying to understand possible limitations of the code

This will help you to get a better understanding of the Python syntax as well as of specific libraries.


## 2. Packages, Libraries, Modules <a class="anchor" id="second-bullet"></a>

One of the key features of Python is that the actual core language is fairly small. This is an intentional design feature to maintain simplicity. Much of the powerful functionality of Python is provided by modules and packages.

We will focus especially on the installation of useful modules for scientific analysis.


### Modules

A module is a file containing Python definitions, functions, and statements. Putting code into modules is useful because of the ability to import the module functionality into your script or IPython session, for instance:






In [7]:
import numpy # good coding style

start = 1
step = 0.5
stop = 100+step #if you want to include the stop

x = numpy.arange(start,stop,step)
print(x)


# x = numpy.arange(start=0,stop=100)


[  1.    1.5   2.    2.5   3.    3.5   4.    4.5   5.    5.5   6.    6.5
   7.    7.5   8.    8.5   9.    9.5  10.   10.5  11.   11.5  12.   12.5
  13.   13.5  14.   14.5  15.   15.5  16.   16.5  17.   17.5  18.   18.5
  19.   19.5  20.   20.5  21.   21.5  22.   22.5  23.   23.5  24.   24.5
  25.   25.5  26.   26.5  27.   27.5  28.   28.5  29.   29.5  30.   30.5
  31.   31.5  32.   32.5  33.   33.5  34.   34.5  35.   35.5  36.   36.5
  37.   37.5  38.   38.5  39.   39.5  40.   40.5  41.   41.5  42.   42.5
  43.   43.5  44.   44.5  45.   45.5  46.   46.5  47.   47.5  48.   48.5
  49.   49.5  50.   50.5  51.   51.5  52.   52.5  53.   53.5  54.   54.5
  55.   55.5  56.   56.5  57.   57.5  58.   58.5  59.   59.5  60.   60.5
  61.   61.5  62.   62.5  63.   63.5  64.   64.5  65.   65.5  66.   66.5
  67.   67.5  68.   68.5  69.   69.5  70.   70.5  71.   71.5  72.   72.5
  73.   73.5  74.   74.5  75.   75.5  76.   76.5  77.   77.5  78.   78.5
  79.   79.5  80.   80.5  81.   81.5  82.   82.5  8

You will find `import` in almost every Python script.

**Question:**
    Importing modules and putting the module name in front is such a bother, why do I need to do this?
    
    
**Answer:**
    It keeps everything modular and separate. For instance many modules have a `read()` function since this is a common thing to do. Without using the `<module>.<function>(...)` syntax there would be no way to know which one to call. 

**Tip:** Sometimes it is convenient to import every function and definition from the module into the current namespace (in other words make them available without using the prefix).


Python allows this with this syntax:

`from module import *`

That means to import every function and definition from the module into the current namespace (in other words make them available without prefixing). For instance you could do:





In [3]:
from numpy import * # no good coding style!
y = arange(100)
print(y)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]


A general rule of thumb is that `from module import * is fine for interactive analysis within IPython but you should avoid using it within scripts where conflicts are more likely.

### Packages

A package is simply a way of collecting related modules together within a single tree-like hierarchy. Very complex packages like NumPy or SciPy have hundreds of individual modules so putting them into a directory-like structure keeps things organized and avoids name collisions. 

**Exercise:** Import a package module and learn about it

Import the Linear algebra module from the SciPy package and find out what functions it provides.



In [9]:

import numpy as np

from scipy import linalg

a = np.array([[1., 2.], [3., 4.]])

b=linalg.inv(a)

c= np.dot(a, b)

print(a)

print(b)

print(c)



[[1. 2.]
 [3. 4.]]
[[-2.   1. ]
 [ 1.5 -0.5]]
[[1.0000000e+00 0.0000000e+00]
 [8.8817842e-16 1.0000000e+00]]


### Finding and installing other packages

Despite we cover here the most common scientific packages used in astronomy, physics and mathematics,
is almost certain that you will eventually find a need that is not met within your current installation. Here we learn where to find other useful packages and how to install them.



**PyPI**

The Python Package Index PyPI is the main repository for 3rd party Python packages (about 14000 packages and growing). An increasing number of astronomy related packages are available on PyPI, but this list misses a lot of available options.

The advantage of being on PyPI is the ease of installation using `pip3 install package_name`


**Exercise:** Find one or more Python packages that will transform coordinates from Galactic to FK5 ecliptic.



**Package installation**

There are two standard methods for installing a package.

The most common one is using `pip3 install`.

The `pip3 install` script is available within our scientific Python installation and is very easy to use (when it works). During the installation process you already saw many examples of pip install in action. Features include:

* If supplied with a package name then it will query the PyPI site to find out about that package. Assuming the package is there then pip install will automatically download and install the package.
* Will accept a local tar file (assuming it contains an installable Python package) or a URL pointing to a tar file.
* Can install in the user package area via `pip install <package or URL> --user` (but see discussion further down)


Another way for installing a package is running an installation script:
`python setup.py install`

**Where do packages get installed?**

An important option in the installation process is where to put the package files. There is a `--user` option in `pip3 install` and `python3 setup.py install`.

Packages get installed in a local user-owned directory when you do something like either of the following:

`pip3 install --user aplp`

`pip install --user aplpy`

`python3 setup.py install --user`


This puts the packages into:

Linux: 	~/.local/lib/python-3.x/site-packages

Mac: 	~/Library/Python/3.x/lib/python/site-packages

Windows: 	%APPDATA%/Python/Python3x/site-packages

**How do I find a package once installed?**

Finding the file associated with a package or module is simple: Just use the `help` command in IPython:

`import scipy`

`help scipy`

This gives something like:


```
NAME
    scipy

FILE
    /usr/local/lib/python2.6/site-packages/scipy/__init__.py

DESCRIPTION
    SciPy: A scientific computing package for Python
    ================================================

    Documentation is available in the docstrings and
    online at http://docs.scipy.org.
```

    
    
    
**Where does Python look for modules?**

The official reference on *Modifying Python's Search Path* gives all the details. In summary:

When the Python interpreter executes an import statement, it looks for modules on a search path. A default value for the path is configured into the Python binary when the interpreter is built. You can determine the path by importing the sys module and printing the value of sys.path:

```
$ python
Python 2.2 (#11, Oct  3 2002, 13:31:27)
[GCC 2.96 20000731 (Red Hat Linux 7.3 2.96-112)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/local/lib/python2.3', '/usr/local/lib/python2.3/plat-linux2',
 '/usr/local/lib/python2.3/lib-tk', '/usr/local/lib/python2.3/lib-dynload',
 '/usr/local/lib/python2.3/site-packages']
>>>
```

Within a script it is possible to adjust the search path by modify sys.path which is just a Python list. Generally speaking you will want to put your path at the front of the list using insert:

```
import sys
sys.path.insert(0, '/my/path/python/packages')
```

You can also add paths to the search path using the PYTHONPATH environment variable.

    
**Uninstalling packages**

There is no simple and fully consistent way to do this unless you use solutions like Anaconda or Canopy (this is why I recommend using Anaconda!). The Python community is working on this one. In most simple cases, however, you can just delete the module file or directory that is revealed by the technique shown above.


## 3. Coding Exercise <a class="anchor" id="third-bullet"></a>

We continue with coding exercises that build upon what we have learned so far.


### 3.1 Problem 1

The phase of a variable star with period P is the decimal part of t/P where t
is the elapsed time since a given date. Calculate the phase when t = 3456.789
days and the period is P = 1.5 days (with phase = 0 at $t_0$ = 0).


### 3.2 Problem 2

Write a script that takes three command line arguments, the first a temperature as a number, the second an input unit (C, K, F) and the third an output
unit (C, K, F). Your script should then convert the input from (Celsius, Kelvin
or Fahrenheit) to the specified output as in the output unit. You can look up online how to use command line arguments.




## Summary <a class="anchor" id="nineth-bullet"></a>

At this point, all of you should have:
* a working Python3 installation, e.g. using the Anaconda/Miniconda Python Distribution
* understanding on how to write simple Python scripts.

Now we're ready to do more serious coding!